src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionParser.cs

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace NuGet.Packaging.Licenses
+{
+    public class LicenseExpressionParser
+    {
+        /// <summary>
+        /// Parses a License Expression if valid.
+        /// The expression would be parsed correct, even if non-standard exceptions are encountered. The non-standard Licenses/Exceptions have metadata on them with which the caller can make decisions.
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns>NuGetLicenseExpression</returns>
+        /// <exception cref="ArgumentException">If the expression has invalid characters.</exception>
+        /// <exception cref="ArgumentException">If the expression itself is invalid. Example: MIT OR OR Apache-2.0, or the MIT or Apache-2.0, because the expressions are case sensitive.</exception>
+        public static NuGetLicenseExpression Parse(string expression)
+        {
+            var tokenizer = new LicenseExpressionTokenizer(expression);
+            if (!tokenizer.HasValidCharacters())
+            {
+                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.NuGetLicenseExpression_InvalidCharacters));
+            }
+
+            return Parse(tokenizer.Tokenize());
+        }
+
+        /// <summary>
+        /// Based on the Shunting Yard algorithm. <see href="https://en.wikipedia.org/wiki/Shunting-yard_algorithm"/>
+        /// Given an array of tokens, generates a NuGetLicenseExpression if valid.
+        /// </summary>
+        /// <param name="infixTokens"></param>
+        /// <returns>NuGetLicenseExpression</returns>
+        /// <exception cref="ArgumentException">If the passed in values are not a valid SPDX License Expression.</exception>
+        private static NuGetLicenseExpression Parse(IEnumerable<LicenseExpressionToken> infixTokens)
+        {
+            var operatorStack = new Stack<LicenseExpressionToken>();
+            var operandStack = new Stack<LicenseExpressionToken>();
+            NuGetLicenseExpression leftHandSideExpression = null;
+            NuGetLicenseExpression rightHandSideExpression = null;
+
+            var lastTokenType = LicenseTokenType.VALUE;
+            var firstPass = true;
+
+            foreach (var token in infixTokens)
+            {
+                var currentTokenType = token.TokenType;
+                switch (token.TokenType)
+                {
+                    case LicenseTokenType.OPENING_BRACKET:
+                        if (!firstPass && !token.TokenType.IsValidPrecedingToken(lastTokenType))
```

Comment:
>IsValidPrecedingToken [](start = 59, length = 21)

What do you think of `IsValidPrecedingToken` accepting a null `LicenseTokenType` to represent that this is the first pass? This may simplify the parsing logic

Reply:
LicenseTokenType is an enum, so usually not a fan of nullable enums. Happy to chat through that though. 

I think it's more readable if the exception is thrown from the parser rather than some generic method.